### ISSUE TITLE
Transactions with specific gas limit are silently failing (status 1, receipt with events, but not on-chain)

### ISSUE THREAD

**Issue by sebastiendan** [0xPolygon/polygon-edge#1372]
# Transactions with specific gas limit are silently failing (status 1, receipt with events, but not on-chain)
## Description
Transactions that are sent with explicit `gasLimit` that is:
- above the lower threshold (threshold under which transactions fail with `intrinsic gas is too low` error)
- below the sufficient gas amount for the transaction

are silently failing. Their receipt has everything that indicates the transaction succeeded (status 1, expected events in the logs, block number, etc.), yet the transaction isn't actually mutating the chain state.

My setup is the following: I'm deploying Axelar's [ConstAddressDeployer contract](https://github.com/axelarnetwork/axelar-gmp-sdk-solidity/blob/main/contracts/deploy/ConstAddressDeployer.sol) that uses `CREATE2` to deploy contracts with constant address. Then, I'm using the deployed `ConstAddressDeployer` contract to deploy Axelar's [TokenDeployer contract](https://github.com/axelarnetwork/axelar-cgp-solidity/blob/main/contracts/TokenDeployer.sol).

When deploying that second contract via a call to the [deploy](https://github.com/axelarnetwork/axelar-gmp-sdk-solidity/blob/main/contracts/deploy/ConstAddressDeployer.sol#L24) function of `ConstAddressDeployer`, I pass a `gasLimit` like below (with `ethers`):

```
const tx = await deployer.deploy(bytecode, salt, {
  gasLimit: <GAS_LIMIT>,
})

await tx.wait()
```

If `GAS_LIMIT` is below what is needed for the transaction to be successfully executed (which can be found by setting an arbitrarily high value and observing the `gasUsed` of the transaction receipt), no error is triggered. When re-sending the same transaction (which should fail because a contract has allegedly been deployed to that same address), the outcome is the same: no error, receipt with status 1 and the expected events, etc.

When tested with a different EVM-based network (e.g., Remix VM), the transaction correctly fails if `GAS_LIMIT` is too low.

The easiest way to reproduce this is to use remix, compile the contracts linked above, and execute the deploy scripts (shared below) with 1. Remix VM and 2. the Injected Provider (MetaMask) connected to a local `polygon-edge` network.

## Your environment
- macOS 13.3.
- `v0.8.1`
- `v0.8.1` tag
- Locally
- Docker-compose (`docker/local`)

## Steps to reproduce
- Run a local `polygon-edge` network with the `docker/local` setup
- Start a Remix session
- Import the two [ConstAddressDeployer contract](https://github.com/axelarnetwork/axelar-gmp-sdk-solidity/blob/main/contracts/deploy/ConstAddressDeployer.sol) and [TokenDeployer contract](https://github.com/axelarnetwork/axelar-cgp-solidity/blob/main/contracts/TokenDeployer.sol) contracts
- Create the two following JavaScript scripts:

<details>
<summary>deployConstAddressDeployer.js</summary>

```js
(async function() {
  try {
    const metadata = JSON.parse(await remix.call('fileManager', 'getFile', 'browser/artifacts/ConstAddressDeployer.json'))
    // the variable web3Provider is a remix global variable object
    const signer = (new ethers.providers.Web3Provider(web3Provider)).getSigner()
    // Create an instance of a Contract Factory
    let factory = new ethers.ContractFactory(metadata.abi, metadata.data.bytecode.object, signer);
    // Notice we pass the constructor's parameters here
    let contract = await factory.deploy();
    // The address the Contract WILL have once mined
    console.log('ConstAddressDeployer', contract.address);

    // The contract is NOT deployed yet; we must wait until it is mined
    await contract.deployed()
    // Done! The contract is deployed.
    console.log('contract deployed')
  } catch (e) {
    console.log(e.message)
  }
})();
```
</details>

<details>
<summary>deployTokenDeployer.js</summary>

```js
(async function() {
  try {
    const constAddressDeployerMetadata = JSON.parse(await remix.call('fileManager', 'getFile', 'browser/artifacts/ConstAddressDeployer.json'))
    const tokenDeployerMetadata = JSON.parse(await remix.call('fileManager', 'getFile', 'browser/artifacts/TokenDeployer.json'))
    // the variable web3Provider is a remix global variable object
    const signer = (new ethers.providers.Web3Provider(web3Provider)).getSigner()

    const CONST_ADDRESS_DEPLOYER_ADDR = "0xbA2a5068c41dfC534150c9dc518959C54C714eB8"

    const deployer = new ethers.Contract(
      CONST_ADDRESS_DEPLOYER_ADDR,
      constAddressDeployerMetadata.abi,
      signer
    )

    const args = []

    let factory = new ethers.ContractFactory(tokenDeployerMetadata.abi, tokenDeployerMetadata.data.bytecode.object, signer);
    const bytecode = factory.getDeployTransaction(...args).data

    const gas = 800_000
    const key = 'ssdkldfsdfsffftf'
    const salt = ethers.utils.keccak256(
      ethers.utils.defaultAbiCoder.encode(['string'], [key.toString()])
    )

    const tx = await deployer
      .deploy(bytecode, salt, {
        gasLimit: BigInt(gas),
      }).catch(error => {
        console.error('error in tx')
      });

    console.log('tx ok')

    const receipt = await tx.wait()
    console.log('OK')
    console.log(receipt)
  } catch (e) {
      console.log('ERROR')
      console.log(e.receipt)
      console.log('gas', gas)
      console.log('gas used', error.receipt.gasUsed.toNumber())
  }
})();
```
</details>

- Select `Remix VM` as the provider
- Run `deployConstAddressDeployer.js` ([Remix guidelines](https://remix-ide.readthedocs.io/en/latest/running_js_scripts.html))
- Run `deployTokenDeployer.js` with a `gas` that is below a couple of millions gas (you can keep the value set in the script)
- Observe a failed transaction
- Update `deployTokenDeployer.js` with a `gas` value above `8_000_000`
- Observe a successful transaction
- Select `Injected Provider (MetaMask)` as the provider
- Run `deployConstAddressDeployer.js`
- Run `deployTokenDeployer.js` with a low `gas` (e.g. initial `800_000` value)
- Observe a dubiously successful transaction
- Re-run the same script with no change and observe again a successful transaction which indicates we were able to deploy again the same bytecode to the same address
- Update `deployTokenDeployer.js` with a `gas` value above `8_000_000`
- Observe a really successful transaction (re-running it fails for the contract was this time actually deployed)

**Comment by sebastiendan** [comment]
Any feedback?

**Comment by rachit77** [comment]
This issue has been fixed in the latest release [v1.0.0-rc1](https://github.com/0xPolygon/polygon-edge/releases/tag/v1.0.0-rc1).
Also, deployTokenDeployer.js had some issues. I've included the updated script [here](https://gist.github.com/rachit77/45fe052a037d99a795864377753f8ba4)
